vulyk/blueprints/gamification/core/rules.py

```diff
@@ -0,0 +1,258 @@
+    included as well as achievement and its badge is.
+    """
+
+    def __init__(self,
+                 id: int,
+                 badge: str,
+                 name: str,
+                 description: str,
+                 bonus: int,
+                 tasks_number: int,
+                 days_number: int,
+                 is_weekend: bool,
+                 is_adjacent: bool) -> None:
+        """
+        :param id: Raw JSON representation hash.
+        :type id: int
+        :param badge: Badge image (either base64 or URL)
+        :type badge: str
+        :param name: Achievement name
+        :type name: str
+        :param description: Achievement textual description
+        :type description: str
+        :param bonus: How much additional points we give for the achievement
+        :type bonus: int
+        :param tasks_number: Number of tasks needed to achieve
+        :type tasks_number: int
+        :param days_number: Number of days needed to achieve
+        :type days_number: int
+        :param is_weekend: Should this achievement be tied up with weekends?
+        :type is_weekend: bool
+        :param is_adjacent: Must it be given for work in adjacent days?
+        :type is_adjacent: bool
+        """
+        self.badge = badge
+        self.name = name
+        self.description = description
+        self.bonus = bonus
+
+        self._tasks_number = tasks_number
+        self._days_number = days_number
+        self._is_weekend = is_weekend
+        self._is_adjacent = is_adjacent
+
+        self._hash = id
+
+        self._validate()
+
+    def _validate(self):
+        """
+        Verifies that the rule complies to internal invariants. Otherwise –
+        the exception must be thrown.
+
+        :raises: RuleValidationException
+        """
+        # no achievements for n tasks closed on m adjacent days/weekends
+        # might be extended later
+        if self._is_adjacent:
+            if self._tasks_number > 0:
+                raise RuleValidationException(
+                    'Counting number of tasks closed in adjacent days/weekends'
+                    ' is not supported.')
+            elif self._days_number == 0:
+
+                raise RuleValidationException(
+                    'Can not set "adjacent" flag with zero days')
+
+        if self._tasks_number == 0 and self._days_number == 0:
+            raise RuleValidationException('There must be at least one '
+                                          'numeric bound')
+
+    @property
+    def id(self):
+        return self._hash
+
+    @property
+    def tasks_number(self) -> int:
+        return self._tasks_number
+
+    @property
+    def days_number(self) -> int:
+        return self._days_number
+
+    @property
+    def is_weekend(self) -> bool:
+        return self._is_weekend
+
+    @property
+    def is_adjacent(self) -> bool:
+        return self._is_adjacent
+
+    @property
+    def limit(self):
```

Comment:
Logic here is not described in docstring.

Also it's quite an obscure property, in different contexts it means different things.